### PR TITLE
Use manif in the contacts component

### PR DESCRIPTION
**Pull request by GiulioRomualdi**
In this PR I removed the `iDyntree` dependency in favor of `manif`.

I also updated the CI adding `manif` dependency. Here I'm cloning manif from   https://github.com/GiulioRomualdi/manif.git because of https://github.com/artivis/manif/issues/156. I already opened a PR upstream https://github.com/artivis/manif/pull/157 as soon as it has been merged I will update this code.

I would like to wait for the review of @S-Dafarra since he is the first developer of this part of the code

### CODE REVIEW

**Comment by prashanthr05** [comment]
> In this PR I removed the iDyntree dependency in favor of manif.

Why was this necessary?

**Comment by GiulioRomualdi** [comment]
https://github.com/dic-iit/bipedal-locomotion-framework/pull/82/commits/7800437e0d94b4a9dc312b5396e821c5d50be7d7 fixes a bug I introduced with this PR. 

If https://github.com/artivis/manif/issues/160 will be acepted
```cpp
leftTransform.coeffs().head<3>() << 0, -0.8, 0;
```
will become 
```cpp
leftTransform.translation() << 0, -0.8, 0;
```

**Comment by GiulioRomualdi** [comment]
> Why was this necessary?

Because of #84 and for the planner of the swing foot I'm implementing right now :smile:

**Comment by prashanthr05** [comment]
> > Why was this necessary?
> 
> Because of #84 and for the planner of the swing foot I'm implementing right now

Great, thanks! Could you add it in the PR description as well?  :D

File: src/Planners/src/TimeVaryingDCMPlanner.cpp
Comment (prashanthr05) [inline src/Planners/src/TimeVaryingDCMPlanner.cpp:347]: You could also use the act() method. It must be doing the same thing.

File: src/Planners/tests/ContactListTest.cpp
Comment (prashanthr05) [inline src/Planners/tests/ContactListTest.cpp:23]: is this a strong check for doubles? Wouldn't you need a check with tolerance as it was offered by iDynTree::checkDoublesAreEqual(a, b)?

File: src/Planners/tests/TimeVaryingDCMPlannerTest.cpp
Comment (prashanthr05) [inline src/Planners/tests/TimeVaryingDCMPlannerTest.cpp:32]: Please add a comment here as to how the coeffs() is ordered for the SE3d object. It should be 'pos+quat'. If someone modifying the tests, mistake it to be something else, then this would result in problems.

File: src/Planners/tests/TimeVaryingDCMPlannerTest.cpp
Comment (prashanthr05) [inline src/Planners/tests/TimeVaryingDCMPlannerTest.cpp:38]: same as above and for other coeff() related lines.

File: src/Planners/tests/TimeVaryingDCMPlannerTest.cpp
Comment (prashanthr05) [reply src/Planners/tests/TimeVaryingDCMPlannerTest.cpp:32]: Typical issue 7800437 fixes a bug I introduced with this PR.

File: src/Planners/tests/ContactListTest.cpp
Comment (S-Dafarra) [reply src/Planners/tests/ContactListTest.cpp:23]: Initially, this was to check if two contacts are *exactly* the same, i.e. two copies of the same object. We could try to have some coherence also with the transformations check. Either they are both checked precisely or with a tolerance.

File: src/Planners/tests/TimeVaryingDCMPlannerTest.cpp
Comment (S-Dafarra) [reply src/Planners/tests/TimeVaryingDCMPlannerTest.cpp:32]: Actually, I would prefer if there is an equivalent of ``setPosition``.

File: src/Planners/tests/TimeVaryingDCMPlannerTest.cpp
Comment (S-Dafarra) [reply src/Planners/tests/TimeVaryingDCMPlannerTest.cpp:38]: Also, shouldn't it be 
```suggestion
    rightTransform.coeffs().head<3>() << 0, 0.8, 0;
```
?

File: src/Planners/tests/TimeVaryingDCMPlannerTest.cpp
Comment (prashanthr05) [reply src/Planners/tests/TimeVaryingDCMPlannerTest.cpp:32]: Relevant issue https://github.com/artivis/manif/issues/160

Comment (S-Dafarra) [reply src/Planners/tests/TimeVaryingDCMPlannerTest.cpp:32]: Thanks for linking the issue, I missed it.

**Comment by GiulioRomualdi** [comment]
@S-Dafarra 

> Looks good to me. The only open points would be:
> 
> * easy translation from `manif::SE3d` to `iDynTree::Transform` objects and vice-versa.

For this one I can create functions for converting the `iDynTree` object to `manif` as far as I've seen there is also the `Eigen::Map` feature in `manif`. Please see [here](https://github.com/artivis/manif/blob/devel/include/manif/impl/se3/SE3_map.h) 

> * is it possible not to use the `coeffs` method? It is really not clear what you are setting, especially for the rotation part.

You are actually right, I don't like to use `coeffs` method because is really error-prone. As @prashanthr05 already pointed in https://github.com/dic-iit/bipedal-locomotion-framework/pull/82#discussion_r471551861, I opened an issue in the main repository https://github.com/artivis/manif/issues/160. Unfortunately, as @artivis wrote in https://github.com/artivis/manif/issues/160#issuecomment-671388708,  the solution I proposed may lead to bugs (e.g. setting a non-unitary quaternion for the rotation)

We can propose some setters like 
```cpp
void quat(const quaternion& q);
void quat(const SO3& SO3);
```

@traversaro @prashanthr05

File: src/Planners/src/TimeVaryingDCMPlanner.cpp
Comment (GiulioRomualdi) [reply src/Planners/src/TimeVaryingDCMPlanner.cpp]: Done https://github.com/dic-iit/bipedal-locomotion-framework/pull/82/commits/74928a3dbf8c3bce47e90b8b87e9e4c11735d161

File: src/Planners/tests/TimeVaryingDCMPlannerTest.cpp
Comment (GiulioRomualdi) [reply src/Planners/tests/TimeVaryingDCMPlannerTest.cpp]: @S-Dafarra 
Fixed in https://github.com/dic-iit/bipedal-locomotion-framework/pull/82/commits/e2e3bc36514b4b761a0ba1d8734473de3694c36a

**Comment by S-Dafarra** [comment]
> For this one I can create functions for converting the `iDynTree` object to `manif` as far as I've seen there is also the `Eigen::Map` feature in `manif`. Please see [here](https://github.com/artivis/manif/blob/devel/include/manif/impl/se3/SE3_map.h)

Sounds good to me.



> You are actually right, I don't like to use `coeffs` method because is really error-prone. As @prashanthr05 already pointed in [#82 (comment)](https://github.com/dic-iit/bipedal-locomotion-framework/pull/82#discussion_r471551861), I opened an issue in the main repository [artivis/manif#160](https://github.com/artivis/manif/issues/160). Unfortunately, as @artivis wrote in [artivis/manif#160 (comment)](https://github.com/artivis/manif/issues/160#issuecomment-671388708), the solution I proposed may lead to bugs (e.g. setting a non-unitary quaternion for the rotation)

I can argue that it is even easier to set non-unitary quaternion with the `coeffs` method 😁 In any case, this is just to have a "clean use-case", a sort of implicit documentation for future usages. Have you tried the SE constructors mentioned in the issue together with some copy assignment?

**Comment by GiulioRomualdi** [comment]
> Have you tried the SE constructors mentioned in the issue together with some copy assignment? 

it should work but I don't like to set both a rotation and a translation if I've to change only one

**Comment by GiulioRomualdi** [comment]
https://github.com/GiulioRomualdi/manif/commit/be76458b7762a8b870d2bf5ca26ec6c9a32d79ed Implements the setters proposed by @artivis

**Comment by GiulioRomualdi** [comment]
In https://github.com/dic-iit/bipedal-locomotion-framework/pull/82/commits/068490e90b04caae42e2f5c2fc82c3fbe254bc33 I avoid to use the `coeffs()`  method in the tests. (it is a workaround for https://github.com/artivis/manif/issues/160) 

Regarding the `Map` is slightly more complex. `Eigen::Map<manif::SE3d>` object wants a contiguous memory 7-d vector containing the translation and the quaternion, on the other hand `iDynTree::Transform` stores the rotation and the translation in two separate objects. What I propose is to create two functions for handling the coping of a `manif` object into an `iDynTree` object and vice-versa. We can discuss this in a separate issue/PR. 

@prashanthr05 @S-Dafarra let me know if it is fine with you so I can merge the PR

**Comment by S-Dafarra** [comment]
> Regarding the `Map` is slightly more complex. `Eigen::Map<manif::SE3d>` object wants a contiguous memory 7-d vector containing the translation and the quaternion, on the other hand `iDynTree::Transform` stores the rotation and the translation in two separate objects. What I propose is to create two functions for handling the coping of a `manif` object into an `iDynTree` object and vice-versa. We can discuss this in a separate issue/PR.
> 
> @prashanthr05 @S-Dafarra let me know if it is fine with you so I can merge the PR

Ok for me

**Review by S-Dafarra — APPROVED** [review]

File: src/Planners/tests/ContactListTest.cpp
Comment (GiulioRomualdi) [reply src/Planners/tests/ContactListTest.cpp:24]: https://github.com/dic-iit/bipedal-locomotion-framework/pull/82/commits/fba1c02cef442dee58e39f09f9eb1aa75cbe2995 goes in the direction of what @prashanthr05 suggested in https://github.com/dic-iit/bipedal-locomotion-framework/pull/82#discussion_r470568742

**Comment by S-Dafarra** [comment]
Even if an approving reviewer is enough, if one of the reviewers requested changes it may be worth waiting for his approval too 😉

**Comment by GiulioRomualdi** [comment]
> Even if an approving reviewer is enough, if one of the reviewers requested changes it may be worth waiting for his approval too

You're right. @prashanthr05  tell me if the change I did are fine with you, if not I open a new PR with the modifications

**Comment by prashanthr05** [comment]
> You're right. @prashanthr05 tell me if the change I did are fine with you, if not I open a new PR with the modifications

:rocket: from me!!

**Pull request merged**
